# Intro to ML

La práctica de hoy consiste en un simulacro de como empezar un proyecto de analítica / Machine Learning desde el principio.

Cómo hemos comentado en clases anteriores, el primer paso, antes incluso que ponerse a analizar el dataset, es plantearse la estrategía a seguir o como plasmar un problema de negocio en un problema de datos.

Nuestro cliente, una importante hotelera mallorquina (oh, sorpresa). No tiene muy claro que es lo que quiere, simplemente sabe que todo el mundo usa IA actualmente y no quiere quedarse atrás. Hablando con ellos, hemos accedido a hacer una propuesta para ayudarles a entender que valor podrían extraer del análisis de datos / modelización. Por el momento, sabemos que están preocupados por la reputación de sus hoteles y que les interesa saber que motiva que un cliente deje una buena review.

Para tal proposito, disponemos de un dataset con reviews de hoteles en Europa. El dataset se encuentra en la carpeta `data` junto con una descripción de los campos disponibles.

## Trabajo por equipos

#### 1. Revisa el fichero `data_format.txt` para saber que variables tenemos disponibles. A partir de las variables, plantea al menos 3 preguntas que te permitan entender el dataset: por ejemplo:

> Cómo varía la puntuación media de un hotel a lo largo del tiempo ?

1. Vemos que tenemos una variable temporal `review_date` disponible. Plantea cómo podríamos transformar esta variable de forma que pueda aportar valor a la modelizción. Piensa que en un contexto de turismo la estacionalidad es un factor muy importante.

1. Vemos que de todas las variables que tenemos, 3 son texto (2 texto libre y 1 tags). Piensa como vas a tratar esas variables. (Nota: muchas veces nos sabremos de antemano cómo tratar una transformación de un tipo de dato. Usar google es legal.)

1. Vemos que como variable numéricas tenemos la Lat/Long. Cómo podríamos usarla ? Que tipo de nuevas features se podrían calcular a partir de ellas que nos aporten informacion ?

1. Si queremos saber que afecta a una review positiva/negativa, como podríamos reducir el problema de ML supervisado ? Preferímos regresion o clasificacion?



## Trabajo entregable

**Ahora que más o menos tenemos una idea inicial de como tratar el dataset...**

NOTA: el objetivo es ver que sabemos usar los pipelines de sklear. No nos volvamos locos con hacer un trabajo completo.

1. Carga las librerías básicas para el analisis de datos
1. Realiza una limpieza no exhaustiva de los datos
1. Separa los datos en train y test. OJO: no queremos filtrar información! Hay que pensar bien como queremos hacer la separación...
1. Plantea una primera aproximación utilizando un modelo lineal y dibuja una arquitectura de cómo quedaría el `pipeline` de transformación + modelización tipo el mostrado en la clase práctiva de la W5
1. Entrena el modelo y revisa las variables/palabras que son sido más importantes.
1. Cuentale al cliente un mensaje clave sobre que tipo de cosas tienen más en cuenta los clientes: `Si no quieres una mala review no sirvas comida fría`

## Resposta

En aquesta llibreta s'apliquen les conclusions i l'après 
al quadern d'[enginyeria de característiques](10_hr_features_engineering.ipynb) 
que s'ha lliurat de forma conjunta amb aquest. 

Així doncs, alguns dels comentaris faran referència al que ja s'ha informat en 
aquell altre quadern de *Jupyter*.

### 1. Carga las librerías básicas para el analisis de datos

#### Llibreries mínimes per càrrega i anàlisi bàsic de les dades

In [1]:
import numpy as np
import pandas as pd

### 2. Realiza una limpieza no exhaustiva de los datos

#### Carregam les dades

In [2]:
data_file_list = [
    'data/hotel_reviews_dataset_part_1.csv',
    'data/hotel_reviews_dataset_part_2.csv', 
    'data/hotel_reviews_dataset_part_3.csv',
    'data/hotel_reviews_dataset_part_4.csv']

df = pd.DataFrame()

for file in data_file_list:
    print(f"Loading {file}")
    df = pd.concat([df, pd.read_csv(file)], ignore_index = True)

Loading data/hotel_reviews_dataset_part_1.csv
Loading data/hotel_reviews_dataset_part_2.csv
Loading data/hotel_reviews_dataset_part_3.csv
Loading data/hotel_reviews_dataset_part_4.csv


In [3]:
df.head(3)

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring Review_Date  Average_Score   Hotel_Name  \
0                           194    8/3/2017            7.7  Hotel Arena   
1                           194    8/3/2017            7.7  Hotel Arena   
2                           194   7/31/2017            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   

                                                Tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
2  [' Leisure trip ', ' Family with young childre...            3 days   

         lat       lng  
0  52.360576  4.915968  
1  52.360576  4.915968  
2  52.360576  4.915968

#### Anàlisi estadístic bàsic

In [4]:
df.describe()

Additional_Number_of_Scoring  Average_Score  \
count                 515738.000000  515738.000000   
mean                     498.081836       8.397487   
std                      500.538467       0.548048   
min                        1.000000       5.200000   
25%                      169.000000       8.100000   
50%                      341.000000       8.400000   
75%                      660.000000       8.800000   
max                     2682.000000       9.800000   

       Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
count                      515738.000000            515738.000000   
mean                           18.539450              2743.743944   
std                            29.690831              2317.464868   
min                             0.000000                43.000000   
25%                             2.000000              1161.000000   
50%                             9.000000              2134.000000   
75%                            23.000000              3613.000000   
max                           408.000000             16670.000000   

       Review_Total_Positive_Word_Counts  \
count                      515738.000000   
mean                           17.776458   
std                            21.804185   
min                             0.000000   
25%                             5.000000   
50%                            11.000000   
75%                            22.000000   
max                           395.000000   

       Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
count                               515738.000000   515738.000000   
mean                                     7.166001        8.395077   
std                                     11.040228        1.637856   
min                                      1.000000        2.500000   
25%                                      1.000000        7.500000   
50%                                      3.000000        8.800000   
75%                                      8.000000        9.600000   
max                                    355.000000       10.000000   

                 lat            lng  
count  512470.000000  512470.000000  
mean       49.442439       2.823803  
std         3.466325       4.579425  
min        41.328376      -0.369758  
25%        48.214662      -0.143372  
50%        51.499981       0.010607  
75%        51.516288       4.834443  
max        52.400181      16.429233

Crida l'atenció que la latitud i la longitud tenguin un recorregut tan petit. 
Però com ja sabem pel treball fet al quadern anterior, 
això és degut a que els hotels són només de 6 ciutats diferents. 

#### Tractament dels valors *nan*

A mode de recordatori del vist al quadern d'enginyeria de característiques, 
miram (per segon cop) si tenim valors *nan* a qualque columna ...

In [5]:
df.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  object 
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  object 
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  object 
 5   Reviewer_Nationality                        515738 non-null  object 
 6   Negative_Review                             515738 non-null  object 
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  object 
 

Pel que fa **als valors nuls** a les coordenades geogràfiques
(camps `lat` i `lng`) d'alguns registres podriem escollir entre: 
- completar el nostre *data set* des de fonts externes, informant els nuls per inferència de les seves coordenades aproximades a partir del país i ciutat o l'adreça completa
- suprimir els registres sense coordenades informades

> Com ja s'ha vist al quadern d'[enginyeria de característiques](10_hr_features_engineering.ipynb#Valors_nuls), el percentatge d'afectació pel que fa als registres és inferior al 0.7%. 
>
> Degut a aquesta escasa afectació, pensam que tenim dues possibilitats: 1) en el cas d'haver d'utilitzar la latitud i longitud com a característiques, **optariem per la supresió dels registres amb nuls**, doncs el més probable és que no valgui la pena l'esforç per completar-les i a més, part de la informació que aportaria la latitud, ja està capturada per la nova característica `hotel_country` vista al quadern d'enginyeria de característiques; 2) si no calés utilitzar les característiques de latitud i longitud, **optariem per utilitzar tot el conjunt de dades** subministrat.

#### Estudi de les característiques predictores i creació de noves

Com ja s'ha dit a l'inici de la [resposta](#Resposta), 
per tal de mantenir aquest *notebook* compacte, 
s'ha optat per fer l'estudi de les característiques en 
el quadern [10_hr_features_engineering.ipynb](10_hr_features_engineering.ipynb) 
que es subministra conjuntament amb aquest.

Per a fer aquesta primera aproximació lineal 
que se'ns demana, optam per a no fer servir 
les característiques de latitud i longitud. 

A mode de recordatori i referència s'inclou la seguent taula resum:

##### Taula 1: resum de la codificació de les característiques

| Columna       | Característica        | Explicació |
| --- | --- | --- |
| Hotel_Address | hotel_id              | Cream un valor de hash únic per a cada establiment. L'utilitzarem per a dividir el conjunt de forma consistent |
| Ídem          | h_country1..5      | Extreim el nom del país de `Hotel_Address` i aplicam *OHE*. Esperam que sigui moderadament explicativa |
| Average_Score | average_score         | És el valor escalat de la columna. Esperam que sigui àltament explicativa |
| Hotel_Name    | Contribueix a hotel_id  | Tot i que el nom de l'hotel podria influenciar a un percentage petit dels hostes, pensam que no val la pena fer-la servir, excepte per a la generació del hash `hotel_id` |
| Total_Number_of_Reviews | hotel_tnreviews  | Conté el valor escalat de la columna. Pensam que un número gran d'opinions pot indicar que l'establiment necessita millorar |
| Additional_Number_of_Scoring | No utilitzada | Pensam que serà molt menys explicativa que `hotel_tnreviewes` i la ignoram. |
| lat      | No utilitzada | Part de la seva informació la capturam amb el país de l'adreça (`hotel_countryN`) |
| lng      | No utilitzada | Ídem |
| Review_Date | rd_2q..4q    | Aquestes variables *dummy* codifiquen el trimestre de l'any en que es grava la valoració. |
| Reviewer_Nationality | rvr_country0..9 | Generam aquestes 10 variables *dummy* aplicant *OHE* als 10 països més freqüents amb l'esperança de que aporti una certa informació respecte de la forma de pensar i valorar de cada nacionalitat. |
| Negative_Review |  | El texte de l'opinió serà valorat un cop entrenat el model |
| Review_Total_Negative_Word_Counts | rv_negwcount | Una opinió negativa més llarga segurament està inversament relacionada amb la nota |
| Positive_Review |  | No utilitzat perquè només ens demanen que ens fixem en el que cal millorar |
| Review_Total_Positive_Word_Counts | rv_poswcount | Una opinió positiva més llarga pot estar (o no) directament relacionada amb una nota més alta |
| Reviewer_Score | y | Formarà el vector de valors a predir |
| Total_Number_of_Reviews_Reviewer_Has_Given | rvr_tnreviews | Pensam que si té experiència en opinar, les opinions seran més extenses a l'hora d'enumerar les falles o les lloances i que la nota molts cops serà proporcional a l'extensió. |
| Tags | Veure taula 2 |  |
| days_since_review | No utilitzada | Pensam que aquesta característica aportarà poca informació o senzillament renou. |

On:
- Columna: nom de la columna original al *data set* subministrat
- Característica: nom de la característica al *data set* processat pel *pipeline*
- Explicació: comentari aclaridor

##### Taula 2: codificació de la característica `Tags`

| Semàntica          | Variable *dummy*     | Val 1 només si existeix l'etiqueta |
| --- | --- | --- |
| tipologia de l'hoste | `gt_fam_old_child`   | `Family with older children` |
| ídem               | `gt_fam_young_child` | `Family with young children` |
| ídem               | `gt_group`           | `Group` |
| ídem               | `gt_solo`            | `Solo traveler` |
| ídem               | `gt_couple`          | `Couple` |
| tipus de viatge    | `tt_business`        | `Business trip` |
| ídem               | `tt_leisure`         | `Leisure trip` |
| durada de l'estada | `stayed_1`           | `Stayed 1 night` |
| ídem               | `stayed_2`           | `Stayed 2 nights` |
| ídem               | `stayed_3`           | `Stayed 3 nights` |
| ídem               | `stayed_4`           | `Stayed 4 nights` |

#### Preparació de la nostra l'arquitectura de *pipelines* 

Seguirem aquestes passes:

##### 1. Definir les classes necessàries

Rescatam les classes SelectColumns i DropColumns del notebook `W5/ 5 - Intro to ML.ipynb`
i afegim les noves que necessitam 
per tractar cada una de transformacions de variables categòriques.

In [6]:
from sklearn.base import TransformerMixin

### aux functions

class SelectColumns(TransformerMixin):
    def __init__(self, columns: list) -> pd.DataFrame:
        if not isinstance(columns, list):
            raise ValueError('Specount the columns into a list')
        self.columns = columns
    def fit(self, X, y=None): # we do not need to specify the target in the transformer. We leave it as optional arg for consistency
        return self
    def transform(self, X):
        return X[self.columns]
    
class DropColumns(TransformerMixin):
    def __init__(self, columns: list) -> pd.DataFrame:
        if not isinstance(columns, list):
            raise ValueError('Specify the columns into a list')
        self.columns = columns
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        return X.drop(self.columns, axis=1)

També afegirem una nova classe amb la qual 
podrem **extreure les dades numériques** 
de columnes amb números, seguides d'unitats, 
com per exemple la `days_since_review` 
si al final decidíssim fer-la servir 
(de moment no la consideram interessant).

In [7]:
class Extract2Num(TransformerMixin):
    def __init__(self, columns: list) -> pd.DataFrame:
        if not isinstance(columns, list):
            raise ValueError('Specify the columns into a list')
        self.columns = columns
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        _X = pd.DataFrame()
        for column in self.columns: 
            _X[column] = X[column].str.extract(r'(\d+)')
        return X.drop(self.columns, axis = 1).join(_X)

La següent **genera un valor hash** a partir d'una tupla 
especificada per `columns`:

In [8]:
import pandas as pd

# afegim el paràmetre rsuffix
# per evitar errors per mor de col·lisions al join final

class HashThis(TransformerMixin):
    def __init__(self, columns: list, 
                 mod: int,
                 new_col: str) -> pd.DataFrame:
        if not isinstance(columns, list):
            raise ValueError('Specify the columns into a list')
        if not isinstance(new_col, str):
            raise ValueError('Specify the new column as str')
        self.columns = columns
        self.mod = mod
        self.new_col = new_col
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        hashes = []
        for t in X[self.columns].itertuples(index = False, name = None):
            hashes.append(hash(str(t)) % self.mod)
        return X.join(
            pd.DataFrame(
                hashes, 
                columns = [self.new_col]).set_index(X.index)
            )

La següent 
**codifica el nom del país en forma de variables dummy**. 
Li passam la columna amb l'adreça a explorar, 
la llista de països i un suffixe per evitar col·lisions 
en fer el *join* amb el *data frame* original:

In [9]:
import pandas as pd

class DummyCountryEnds(TransformerMixin):
    def _zero_1(self, length: int, one_pos: int) -> list:
        """
        Return zero'ed vector of length, except at one_pos
        """
        v = [0] * length
        v[one_pos] = 1
        return v
    
    def __init__(self, column: str, 
                 country_list: list,
                 prefix: str
                ) -> pd.DataFrame:
        if not isinstance(column, str):
            raise ValueError('Specify the column as str')
        if not isinstance(country_list, list):
            raise ValueError('Specify the country_list as a list')
        if not isinstance(prefix, str):
            raise ValueError('Specify the columns name prefix as str')
        self.column = column
        self.country_list = country_list
        self.prefix = prefix
        self.list_len = len(country_list)
        # construir llista de noms de columna
        self.columns = list(prefix + str(i) for i in range(self.list_len))
        # construir diccionari de dummys
        self.country_dict = {
            country_list[i]: self._zero_1(self.list_len, i) 
            for i in range(0, self.list_len)}
        self.country_dict['None'] = [0] * self.list_len
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        country_dummy = []
        for address in X[self.column]:
            country_name = 'None'
            i = 0
            while (country_name == 'None') and (i < self.list_len):
                if address.endswith(self.country_list[i]):
                    country_name = self.country_list[i]
                i += 1
            country_dummy.append(self.country_dict[country_name])
        return X.join(
            pd.DataFrame(
                country_dummy, 
                columns = self.columns).set_index(X.index)
            )

La proper també 
**codifica el nom del país en forma de variables dummy**, 
però ho fa comparant tota la variable. 
Li passam la columna amb la'dreça a explorar, 
la llista de països i un suffixe per evitar col·lisions 
en fer el *join* amb el *data frame* original:

In [10]:
import pandas as pd

class DummyCountryIs(TransformerMixin):
    def _zero_1(self, length: int, one_pos: int) -> list:
        """
        Return zero'ed vector of length, except at one_pos
        """
        v = [0] * length
        v[one_pos] = 1
        return v
    
    def __init__(self, column: str, 
                 country_list: list,
                 prefix: str
                ) -> pd.DataFrame:
        if not isinstance(column, str):
            raise ValueError('Specify the column as str')
        if not isinstance(country_list, list):
            raise ValueError('Specify the country_list as a list')
        if not isinstance(prefix, str):
            raise ValueError('Specify the columns name prefix as str')
        self.column = column
        self.country_list = country_list
        self.prefix = prefix
        self.list_len = len(country_list)
        # construir llista de noms de columna
        self.columns = list(prefix + str(i) for i in range(self.list_len))
        # construir diccionari de dummys
        self.country_dict = {
            country_list[i]: self._zero_1(self.list_len, i) 
            for i in range(0, self.list_len)}
        self.country_dict['None'] = [0] * self.list_len
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        country_dummy = []
        for country in X[self.column]:
            country_name = 'None'
            i = 0
            while (country_name == 'None') and (i < self.list_len):
                if country.strip() == self.country_list[i]:
                    country_name = self.country_list[i]
                i += 1
            country_dummy.append(self.country_dict[country_name])
        return X.join(
            pd.DataFrame(
                country_dummy, 
                columns = self.columns).set_index(X.index)
            )

Amb la que segueix, cream les **variables *dummy* del trimestre** 
de la data de la gravació.

In [11]:
import pandas as pd
from datetime import datetime

class DummyDateQuarter(TransformerMixin):
    def _zero_1(self, length: int, one_pos: int) -> list:
        """
        Return zero'ed vector of length, except at one_pos
        """
        v = [0] * length
        v[one_pos] = 1
        return v
    
    def __init__(self, column: str, 
                 dformat: str, 
                 prefix: str
                ) -> pd.DataFrame:
        if not isinstance(column, str):
            raise ValueError('Specify the column as str')
        if not isinstance(dformat, str):
            raise ValueError('Specify the date format as datetime format string')
        if not isinstance(prefix, str):
            raise ValueError('Specify the columns name prefix as str')
        self.column = column
        self.dformat = dformat
        self.prefix = prefix
        # construir llista de noms de columna
        self.columns = list(prefix + str(i) for i in range(4))
        # construir diccionari de dummys
        self.month_dict = {
            i + 1: self._zero_1(4, i // 3) 
            for i in range(12)}
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        quarter_dummy = []
        for date in X[self.column]:
            quarter_dummy.append(
                self.month_dict[
                    datetime.strptime(date, self.dformat).month
                    ]
                )
        return X.join(
            pd.DataFrame(
                quarter_dummy, 
                columns = self.columns).set_index(X.index)
            )

I finalment **processam les etiquetes**, convertint-les en variables *dummy*.

In [12]:
import pandas as pd
from ast import literal_eval

class DummyTags(TransformerMixin):    
    def __init__(self, column: str, 
                 tags_list: list,
                 prefix: str
                ) -> pd.DataFrame:
        if not isinstance(column, str):
            raise ValueError('Specify the column as str')
        if not isinstance(tags_list, list):
            raise ValueError('Specify the country_list as a list')
        if not isinstance(prefix, str):
            raise ValueError('Specify the columns name prefix as str')
        self.column = column
        self.tags_list = tags_list
        self.prefix = prefix
        self.list_len = len(tags_list)
        # construir llista de noms de columna
        self.columns = list(prefix + str(i) for i in range(self.list_len))
        self.zeros = [0] * self.list_len
    def fit(self, X, y=None):
        return self
    def transform(self, X):
        tags_dummy = []
        for tags in X[self.column]:
            tags = literal_eval(tags)
            dummies = [0] * self.list_len
            for i in range(self.list_len):
                if self.tags_list[i] in tags:
                    dummies[i] = 1
            tags_dummy.append(dummies)
        return X.join(
            pd.DataFrame(
                tags_dummy, 
                columns = self.columns).set_index(X.index)
            )

##### 2. Segregar les columnes per itineraris

Separarem les columnes en quatre classes:
- les numèriques que necessiten netejar valors *nan* i després ésser escalades
- les categòriques, que requereixen d'una transformació (per exemple, codificar per mitjà de variables *dummy*)
- les que cal suprimir al final del procés de transformació i abans d'aplicar-les al model
- el valor que volem predir

In [13]:
from sklearn.pipeline import Pipeline

In [14]:
num_cols_list = [
    'Average_Score',
    'Total_Number_of_Reviews',
    'Review_Total_Negative_Word_Counts',
    'Review_Total_Positive_Word_Counts',
    'Total_Number_of_Reviews_Reviewer_Has_Given',
    'Reviewer_Score']

cat_cols_list = [
    'Hotel_Address', 
    'Hotel_Name',
    'Review_Date',
    'Reviewer_Nationality',
    'Tags']

# eliminam també una columna del país de l'hotel 
# i una del trimestre per evitar colinealitats
drop_cat_cols_list = [
    'Hotel_Address',
    'Hotel_Name',
    'Review_Date',
    'Reviewer_Nationality',
    'Tags',
    'hc0', # primera columna del país (OHE)
    'rd_q0' # primera columna del trimestre
    ]

Itinerari columnes numèriques:

In [15]:
from sklearn.preprocessing import MinMaxScaler

sel_num_col_step = ('sel_num_col', 
                    SelectColumns(num_cols_list)
                   )

scaler_step = ('scaler', 
               MinMaxScaler()
              )

num_pipe_steps = [sel_num_col_step, scaler_step]

num_pipe = Pipeline(num_pipe_steps)

Itinerari columnes categòriques:

In [16]:
# Columnes per a la generació del codi únic d'establiment hoteler
hotel_id_cols = ['Hotel_Address', 'Hotel_Name']
hash_mod = 1e7 - 1

# Extracció del país des de l'adreça de l'hotel
hotel_address_col = 'Hotel_Address'
hotel_country_list = ['Austria',
                      'France', 
                      'Italy',
                      'Netherlands', 
                      'Spain', 
                      'United Kingdom']

# Extracció del trimestre
rv_date_col = 'Review_Date'
rv_date_format = '%m/%d/%Y'

# Llista de les 10 nacionalitats més freqüents
rvr_country_col = 'Reviewer_Nationality'
rvr_country_list = ['Australia',
                    'Canada', 
                    'Germany', 
                    'Ireland',
                    'Netherlands', 
                    'Saudi Arabia', 
                    'Switzerland', 
                    'United Arab Emirates',
                    'United Kingdom',
                    'United States of America']

# Extracció des d'etiquetes
rv_tags_col = 'Tags'
rv_tags_list = [' Family with older children ',
                ' Family with young children ',
                ' Group ',
                ' Solo traveler ',
                ' Couple ',
                ' Business trip ',
                ' Leisure trip ',
                ' Stayed 1 night ',
                ' Stayed 2 nights ',
                ' Stayed 3 nights ',
                ' Stayed 4 nights ']

# Que volem predir?

src_target_col = 'Reviewer_Score'
target_col = 'y'

In [17]:
# farem servir OneHotEncoder a mida per tal de garantir 
# que els camps dummy sempre facin referència al mateix

sel_cat_col_step = ('sel_cat_col', 
                    SelectColumns(cat_cols_list)
                   )

hotel_id_step = ('hotel_id', 
                 HashThis(columns = hotel_id_cols, 
                          mod = hash_mod,
                          new_col = 'hotel_id')
                )

hotel_country_step = ('hotel_country', 
                      DummyCountryEnds(column = hotel_address_col, 
                                       country_list = hotel_country_list,
                                       prefix = 'hc')
                     )

rv_date_step = ('rd_q',
                DummyDateQuarter(column = rv_date_col, 
                                 dformat = rv_date_format,
                                 prefix = 'rd_q')
               )

rvr_country_step = ('rvr_country', 
                    DummyCountryIs(column = rvr_country_col, 
                                   country_list = rvr_country_list,
                                   prefix = 'rc')
                   )

rv_tags_step = ('rv_tags', 
                    DummyTags(column = rv_tags_col, 
                              tags_list = rv_tags_list,
                              prefix = 'tag')
                   )

drop_cat_col_step = ('drop_cat_col', DropColumns(drop_cat_cols_list))

cat_pipe_steps = [sel_cat_col_step,
                  hotel_id_step,
                  hotel_country_step,
                  rv_date_step,
                  rvr_country_step, 
                  rv_tags_step,
                  drop_cat_col_step]

cat_pipe = Pipeline(cat_pipe_steps)

Després les tornam a unir per a obtenir les tuples a processar

In [18]:
from sklearn.pipeline import FeatureUnion

transformer_list = [('cat_pipe', cat_pipe),
                    ('num_pipe', num_pipe)]

### Juntam els dos itineraris 

data_prep_pipe = FeatureUnion(transformer_list=transformer_list)

data_prep_step = ('data_prep', data_prep_pipe)

In [19]:
out_cols = ['hotel_id',
#            'h_country0', # colinealitat
            'h_country1', 'h_country2', 
            'h_country3', 'h_country4', 'h_country5',
#            'rd_1q', # colinealitat
            'rd_2q', 'rd_3q', 'rd_4q',
            'rvr_country0', 'rvr_country1', 
            'rvr_country2', 'rvr_country3', 
            'rvr_country4', 'rvr_country5', 
            'rvr_country6', 'rvr_country7', 
            'rvr_country8', 'rvr_country9', 
            'gt_fam_old_child', 'gt_fam_young_child',
            'gt_group', 'gt_solo', 'gt_couple',      
            'tt_business', 'tt_leisure',
            'stayed_1', 'stayed_2', 'stayed_3', 'stayed_4',
            'average_score', 
            'hotel_tnreviews', 
            'rv_negwcount', 
            'rv_poswcount', 
            'rvr_tnreviews',
            'y']

#### Prova del pipeline

Provam a veure que tal ha quedat:

In [20]:
main_pipe = Pipeline([data_prep_step])

In [21]:
X_pipe = pd.DataFrame(
    main_pipe.fit_transform(df), columns = out_cols)

In [22]:
X_pipe.describe()

hotel_id     h_country1     h_country2     h_country3  \
count  5.157380e+05  515738.000000  515738.000000  515738.000000   
mean   5.056472e+06       0.116199       0.072143       0.110936   
std    2.789716e+06       0.320463       0.258725       0.314053   
min    3.830000e+03       0.000000       0.000000       0.000000   
25%    2.858406e+06       0.000000       0.000000       0.000000   
50%    5.065265e+06       0.000000       0.000000       0.000000   
75%    7.330856e+06       0.000000       0.000000       0.000000   
max    9.990577e+06       1.000000       1.000000       1.000000   

          h_country4     h_country5          rd_2q          rd_3q  \
count  515738.000000  515738.000000  515738.000000  515738.000000   
mean        0.116627       0.508594       0.256465       0.276144   
std         0.320976       0.499927       0.436682       0.447089   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         0.000000       1.000000       1.000000       1.000000   
max         1.000000       1.000000       1.000000       1.000000   

               rd_4q   rvr_country0  ...       stayed_1      stayed_2  \
count  515738.000000  515738.000000  ...  515738.000000  515738.00000   
mean        0.231552       0.042048  ...       0.375472       0.25970   
std         0.421824       0.200700  ...       0.484245       0.43847   
min         0.000000       0.000000  ...       0.000000       0.00000   
25%         0.000000       0.000000  ...       0.000000       0.00000   
50%         0.000000       0.000000  ...       0.000000       0.00000   
75%         0.000000       0.000000  ...       1.000000       1.00000   
max         1.000000       1.000000  ...       1.000000       1.00000   

            stayed_3       stayed_4  average_score  hotel_tnreviews  \
count  515738.000000  515738.000000  515738.000000    515738.000000   
mean        0.185794       0.092716       0.695106         0.162431   
std         0.388941       0.290034       0.119141         0.139380   
min         0.000000       0.000000       0.000000         0.000000   
25%         0.000000       0.000000       0.630435         0.067240   
50%         0.000000       0.000000       0.695652         0.125759   
75%         0.000000       0.000000       0.782609         0.214711   
max         1.000000       1.000000       1.000000         1.000000   

        rv_negwcount   rv_poswcount  rvr_tnreviews              y  
count  515738.000000  515738.000000  515738.000000  515738.000000  
mean        0.045440       0.045004       0.017418       0.786010  
std         0.072772       0.055200       0.031187       0.218381  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.004902       0.012658       0.000000       0.666667  
50%         0.022059       0.027848       0.005650       0.840000  
75%         0.056373       0.055696       0.019774       0.946667  
max         1.000000       1.000000       1.000000       1.000000  

[8 rows x 36 columns]

Comprovem si hi ha qualque NaN:

In [23]:
X_pipe.isnull().values.any()

False

I el conjunt de valors

In [24]:
X_pipe.nunique()

hotel_id              1494
h_country1               2
h_country2               2
h_country3               2
h_country4               2
h_country5               2
rd_2q                    2
rd_3q                    2
rd_4q                    2
rvr_country0             2
rvr_country1             2
rvr_country2             2
rvr_country3             2
rvr_country4             2
rvr_country5             2
rvr_country6             2
rvr_country7             2
rvr_country8             2
rvr_country9             2
gt_fam_old_child         2
gt_fam_young_child       2
gt_group                 2
gt_solo                  2
gt_couple                2
tt_business              2
tt_leisure               2
stayed_1                 2
stayed_2                 2
stayed_3                 2
stayed_4                 2
average_score           34
hotel_tnreviews       1142
rv_negwcount           402
rv_poswcount           365
rvr_tnreviews          198
y                       37
dtype: int64

Amb el rang de *hash* que hem fixat seguim tenguent els 1494 establiments, per tant no hi ha col·lisions.

Inspeccionem breument les columnes:

In [25]:
X_pipe.iloc[:, :9]

hotel_id  h_country1  h_country2  h_country3  h_country4  h_country5  \
0       2071996.0         0.0         0.0         1.0         0.0         0.0   
1       2071996.0         0.0         0.0         1.0         0.0         0.0   
2       2071996.0         0.0         0.0         1.0         0.0         0.0   
3       2071996.0         0.0         0.0         1.0         0.0         0.0   
4       2071996.0         0.0         0.0         1.0         0.0         0.0   
...           ...         ...         ...         ...         ...         ...   
515733  1883384.0         0.0         0.0         0.0         0.0         0.0   
515734  1883384.0         0.0         0.0         0.0         0.0         0.0   
515735  1883384.0         0.0         0.0         0.0         0.0         0.0   
515736  1883384.0         0.0         0.0         0.0         0.0         0.0   
515737  1883384.0         0.0         0.0         0.0         0.0         0.0   

        rd_2q  rd_3q  rd_4q  
0         0.0    1.0    0.0  
1         0.0    1.0    0.0  
2         0.0    1.0    0.0  
3         0.0    1.0    0.0  
4         0.0    1.0    0.0  
...       ...    ...    ...  
515733    0.0    1.0    0.0  
515734    0.0    1.0    0.0  
515735    0.0    1.0    0.0  
515736    0.0    1.0    0.0  
515737    0.0    1.0    0.0  

[515738 rows x 9 columns]

In [26]:
X_pipe.iloc[:, 9:19]

rvr_country0  rvr_country1  rvr_country2  rvr_country3  rvr_country4  \
0                0.0           0.0           0.0           0.0           0.0   
1                0.0           0.0           0.0           1.0           0.0   
2                1.0           0.0           0.0           0.0           0.0   
3                0.0           0.0           0.0           0.0           0.0   
4                0.0           0.0           0.0           0.0           0.0   
...              ...           ...           ...           ...           ...   
515733           0.0           0.0           0.0           0.0           0.0   
515734           0.0           0.0           0.0           0.0           0.0   
515735           0.0           0.0           0.0           0.0           0.0   
515736           0.0           0.0           0.0           0.0           0.0   
515737           0.0           0.0           0.0           0.0           0.0   

        rvr_country5  rvr_country6  rvr_country7  rvr_country8  rvr_country9  
0                0.0           0.0           0.0           0.0           0.0  
1                0.0           0.0           0.0           0.0           0.0  
2                0.0           0.0           0.0           0.0           0.0  
3                0.0           0.0           0.0           1.0           0.0  
4                0.0           0.0           0.0           0.0           0.0  
...              ...           ...           ...           ...           ...  
515733           0.0           0.0           0.0           0.0           0.0  
515734           0.0           0.0           0.0           0.0           0.0  
515735           0.0           0.0           0.0           0.0           0.0  
515736           0.0           0.0           0.0           0.0           0.0  
515737           0.0           0.0           0.0           0.0           0.0  

[515738 rows x 10 columns]

In [44]:
X_pipe.iloc[:, 19:30]

gt_fam_old_child  gt_fam_young_child  gt_group  gt_solo  gt_couple  \
0                    0.0                 0.0       0.0      0.0        1.0   
1                    0.0                 0.0       0.0      0.0        1.0   
2                    0.0                 1.0       0.0      0.0        0.0   
3                    0.0                 0.0       0.0      1.0        0.0   
4                    0.0                 0.0       0.0      0.0        1.0   
...                  ...                 ...       ...      ...        ...   
515733               1.0                 0.0       0.0      0.0        0.0   
515734               0.0                 1.0       0.0      0.0        0.0   
515735               1.0                 0.0       0.0      0.0        0.0   
515736               0.0                 0.0       1.0      0.0        0.0   
515737               0.0                 1.0       0.0      0.0        0.0   

        tt_business  tt_leisure  stayed_1  stayed_2  stayed_3  stayed_4  
0               0.0         1.0       0.0       0.0       0.0       0.0  
1               0.0         1.0       0.0       0.0       0.0       1.0  
2               0.0         1.0       0.0       0.0       1.0       0.0  
3               0.0         1.0       0.0       0.0       1.0       0.0  
4               0.0         1.0       0.0       1.0       0.0       0.0  
...             ...         ...       ...       ...       ...       ...  
515733          0.0         1.0       0.0       0.0       0.0       0.0  
515734          0.0         1.0       0.0       1.0       0.0       0.0  
515735          0.0         1.0       0.0       0.0       0.0       1.0  
515736          0.0         1.0       0.0       0.0       1.0       0.0  
515737          0.0         1.0       0.0       1.0       0.0       0.0  

[515738 rows x 11 columns]

In [28]:
X_pipe.iloc[:, 30:]

average_score  hotel_tnreviews  rv_negwcount  rv_poswcount  \
0            0.543478         0.081795      0.973039      0.027848   
1            0.543478         0.081795      0.000000      0.265823   
2            0.543478         0.081795      0.102941      0.053165   
3            0.543478         0.081795      0.514706      0.065823   
4            0.543478         0.081795      0.343137      0.020253   
...               ...              ...           ...           ...   
515733       0.630435         0.167198      0.034314      0.005063   
515734       0.630435         0.167198      0.026961      0.027848   
515735       0.630435         0.167198      0.046569      0.000000   
515736       0.630435         0.167198      0.000000      0.063291   
515737       0.630435         0.167198      0.031863      0.015190   

        rvr_tnreviews         y  
0            0.016949  0.053333  
1            0.016949  0.666667  
2            0.022599  0.613333  
3            0.000000  0.173333  
4            0.005650  0.560000  
...               ...       ...  
515733       0.019774  0.600000  
515734       0.031073  0.440000  
515735       0.005650  0.000000  
515736       0.005650  0.840000  
515737       0.000000  0.773333  

[515738 rows x 6 columns]

Sembla tot correcte. :-)

### 3. Separa los datos en train y test. 
OJO: no queremos filtrar información! Hay que pensar bien como queremos hacer la separación...

In [29]:
# Reservarem aquest tant per ú per a prova

test_size = 0.15

Com que no volem filtració d'informació, 
hem de mantenir les opinions de cada hotel juntes. 
És a dir cada establiment i per tant 
el conjunt de les opinions que l'afecten
només pot estar al conjunt d'entrenament 
o al de prova.

Per a separar per grups, en lloc de fer servir 
la classe
`GroupKFold`
de sklearn, provarem el mètode de la ruleta russa. 
Ho feim així com a experiment 
i també, perquè no dir-ho, 
no hem sabut de la classe `GroupKFold` fins després
de la nostra implementació:

In [30]:
# Ho feim reproduible per poder depurar millor

np.random.seed(42)

# Feim partició train | test

## Farem servir una ruleta rusa .. comptam quan hi ha de cada hotel

roulette = pd.DataFrame(df[hotel_id_cols].value_counts(hotel_id_cols))
roulette.rename(columns = {0:'cardinality'}, inplace = True)

## Ara remenam una estona

roulette = roulette.sample(frac = 1)

## I cream una màscara per classificar entre train i test

nrecs = roulette['cardinality'].sum()

ntests = 0
to = 0

while (ntests < test_size * nrecs) and (to < len(roulette)):
    ntests += roulette.iloc[to]['cardinality']
    to += 1

print(f"Han estat seleccionats:\n\n"
      f" {ntests} registres"
      f" (el {100 * ntests / nrecs:.2f}%) per a test i\n"\
      f" {nrecs - ntests} registres"
      f" per a entrenament\n\n"\
      f" {nrecs} és el total de registres."
     )

Han estat seleccionats:

 78168 registres (el 15.16%) per a test i
 437570 registres per a entrenament

 515738 és el total de registres.


In [31]:
X_train = df.join(
    roulette.drop('cardinality', axis = 1
                 ).iloc[to:], on = hotel_id_cols, 
    how = 'inner')

X_test  = df.join(
    roulette.drop('cardinality', axis = 1).iloc[:to], 
    on = hotel_id_cols, how = 'inner')

Comprovam que no hi hagi nuls en lloc, excepte a les 
columnes de latitud i longitud.

In [32]:
X_train.iloc[:, :-2].isnull().values.any()

False

In [33]:
X_train[src_target_col].describe()

count    437570.000000
mean          8.387980
std           1.642856
min           2.500000
25%           7.500000
50%           8.800000
75%           9.600000
max          10.000000
Name: Reviewer_Score, dtype: float64

In [34]:
X_test[src_target_col].describe()

count    78168.000000
mean         8.434801
std          1.609011
min          2.500000
25%          7.500000
50%          8.800000
75%          9.600000
max         10.000000
Name: Reviewer_Score, dtype: float64

La mitjana i desviació típica de la variable a predir 
tenen valors semblants a ambdós conjunts, 
segurament per mor de la gran quantitat de mostres.
Per tant no cal fer-ne cap consideració en quan 
a la seva estratificació i podem donar el nostre 
algorisme de partició per correcte.

In [35]:
# Calculam ja els conjunts transformats

## Conjunt d'entrenament
X_train_trans = pd.DataFrame(
    main_pipe.fit_transform(X_train), columns = out_cols)

y_train_trans = X_train_trans[target_col]
X_train_trans.drop(target_col, axis = 1, inplace = True)

## Conjunt de prova
X_test_trans = pd.DataFrame(
    main_pipe.transform(X_test), columns = out_cols)

y_test_trans = X_test_trans[target_col]
X_test_trans.drop(target_col, axis = 1, inplace = True)

Comprovam que tampoc hi hagi nuls a la transformada.

In [36]:
X_train_trans.isnull().values.any()

False

In [37]:
y_train_trans.isnull().values.any()

False

In [38]:
X_test_trans.isnull().values.any()

False

In [39]:
y_test_trans.isnull().values.any()

False

Sembla que no hi ha cap valor nul.
Anem a veure que hi ha:

In [40]:
X_train_trans

hotel_id  h_country1  h_country2  h_country3  h_country4  h_country5  \
0       2071996.0         0.0         0.0         1.0         0.0         0.0   
1       2071996.0         0.0         0.0         1.0         0.0         0.0   
2       2071996.0         0.0         0.0         1.0         0.0         0.0   
3       2071996.0         0.0         0.0         1.0         0.0         0.0   
4       2071996.0         0.0         0.0         1.0         0.0         0.0   
...           ...         ...         ...         ...         ...         ...   
437565  1883384.0         0.0         0.0         0.0         0.0         0.0   
437566  1883384.0         0.0         0.0         0.0         0.0         0.0   
437567  1883384.0         0.0         0.0         0.0         0.0         0.0   
437568  1883384.0         0.0         0.0         0.0         0.0         0.0   
437569  1883384.0         0.0         0.0         0.0         0.0         0.0   

        rd_2q  rd_3q  rd_4q  rvr_country0  ...  tt_leisure  stayed_1  \
0         0.0    1.0    0.0           0.0  ...         1.0       0.0   
1         0.0    1.0    0.0           0.0  ...         1.0       0.0   
2         0.0    1.0    0.0           1.0  ...         1.0       0.0   
3         0.0    1.0    0.0           0.0  ...         1.0       0.0   
4         0.0    1.0    0.0           0.0  ...         1.0       0.0   
...       ...    ...    ...           ...  ...         ...       ...   
437565    0.0    1.0    0.0           0.0  ...         1.0       0.0   
437566    0.0    1.0    0.0           0.0  ...         1.0       0.0   
437567    0.0    1.0    0.0           0.0  ...         1.0       0.0   
437568    0.0    1.0    0.0           0.0  ...         1.0       0.0   
437569    0.0    1.0    0.0           0.0  ...         1.0       0.0   

        stayed_2  stayed_3  stayed_4  average_score  hotel_tnreviews  \
0            0.0       0.0       0.0       0.543478         0.081795   
1            0.0       0.0       1.0       0.543478         0.081795   
2            0.0       1.0       0.0       0.543478         0.081795   
3            0.0       1.0       0.0       0.543478         0.081795   
4            1.0       0.0       0.0       0.543478         0.081795   
...          ...       ...       ...            ...              ...   
437565       0.0       0.0       0.0       0.630435         0.167198   
437566       1.0       0.0       0.0       0.630435         0.167198   
437567       0.0       0.0       1.0       0.630435         0.167198   
437568       0.0       1.0       0.0       0.630435         0.167198   
437569       1.0       0.0       0.0       0.630435         0.167198   

        rv_negwcount  rv_poswcount  rvr_tnreviews  
0           0.973039      0.027848       0.018237  
1           0.000000      0.265823       0.018237  
2           0.102941      0.053165       0.024316  
3           0.514706      0.065823       0.000000  
4           0.343137      0.020253       0.006079  
...              ...           ...            ...  
437565      0.034314      0.005063       0.021277  
437566      0.026961      0.027848       0.033435  
437567      0.046569      0.000000       0.006079  
437568      0.000000      0.063291       0.006079  
437569      0.031863      0.015190       0.000000  

[437570 rows x 35 columns]

In [41]:
y_train_trans

0         0.053333
1         0.666667
2         0.613333
3         0.173333
4         0.560000
            ...   
437565    0.600000
437566    0.440000
437567    0.000000
437568    0.840000
437569    0.773333
Name: y, Length: 437570, dtype: float64

In [42]:
X_test_trans

hotel_id  h_country1  h_country2  h_country3  h_country4  h_country5  \
0      9658097.0         1.0         0.0         0.0         0.0         0.0   
1      9658097.0         1.0         0.0         0.0         0.0         0.0   
2      9658097.0         1.0         0.0         0.0         0.0         0.0   
3      9658097.0         1.0         0.0         0.0         0.0         0.0   
4      9658097.0         1.0         0.0         0.0         0.0         0.0   
...          ...         ...         ...         ...         ...         ...   
78163  4795403.0         0.0         0.0         0.0         0.0         1.0   
78164  4795403.0         0.0         0.0         0.0         0.0         1.0   
78165  4795403.0         0.0         0.0         0.0         0.0         1.0   
78166  4795403.0         0.0         0.0         0.0         0.0         1.0   
78167  4795403.0         0.0         0.0         0.0         0.0         1.0   

       rd_2q  rd_3q  rd_4q  rvr_country0  ...  tt_leisure  stayed_1  stayed_2  \
0        0.0    1.0    0.0           0.0  ...         1.0       0.0       0.0   
1        0.0    1.0    0.0           0.0  ...         1.0       0.0       1.0   
2        0.0    1.0    0.0           0.0  ...         1.0       0.0       0.0   
3        1.0    0.0    0.0           0.0  ...         1.0       0.0       1.0   
4        1.0    0.0    0.0           1.0  ...         1.0       0.0       1.0   
...      ...    ...    ...           ...  ...         ...       ...       ...   
78163    0.0    0.0    1.0           0.0  ...         1.0       1.0       0.0   
78164    0.0    0.0    1.0           0.0  ...         1.0       1.0       0.0   
78165    0.0    0.0    1.0           0.0  ...         1.0       0.0       0.0   
78166    0.0    0.0    1.0           0.0  ...         0.0       0.0       0.0   
78167    0.0    0.0    1.0           0.0  ...         1.0       1.0       0.0   

       stayed_3  stayed_4  average_score  hotel_tnreviews  rv_negwcount  \
0           1.0       0.0       0.739130         0.019727      0.083333   
1           0.0       0.0       0.739130         0.019727      0.083333   
2           0.0       1.0       0.739130         0.019727      0.053922   
3           0.0       0.0       0.739130         0.019727      0.159314   
4           0.0       0.0       0.739130         0.019727      0.004902   
...         ...       ...            ...              ...           ...   
78163       0.0       0.0       0.630435         0.058640      0.000000   
78164       0.0       0.0       0.630435         0.058640      0.041667   
78165       1.0       0.0       0.630435         0.058640      0.000000   
78166       1.0       0.0       0.630435         0.058640      0.000000   
78167       0.0       0.0       0.630435         0.058640      0.000000   

       rv_poswcount  rvr_tnreviews  
0          0.030380       0.015198  
1          0.420253       0.003040  
2          0.030380       0.039514  
3          0.022785       0.018237  
4          0.118987       0.009119  
...             ...            ...  
78163      0.015190       0.003040  
78164      0.027848       0.000000  
78165      0.068354       0.003040  
78166      0.045570       0.024316  
78167      0.048101       0.000000  

[78168 rows x 35 columns]

In [43]:
y_test_trans

0        0.506667
1        0.773333
2        0.666667
3        0.613333
4        1.000000
           ...   
78163    0.946667
78164    0.386667
78165    1.000000
78166    1.000000
78167    1.000000
Name: y, Length: 78168, dtype: float64

### 4. Plantea una primera aproximación utilizando un modelo lineal ... 
... y dibuja una arquitectura de cómo quedaría el `pipeline` de transformación + modelización tipo el mostrado en la clase práctiva de la W5

In [51]:
from sklearn.ensemble import RandomForestRegressor

reg_model = RandomForestRegressor(n_estimators = 10, random_state = 42)

reg_model.fit(X_train_trans, y_train_trans)

RandomForestRegressor(n_estimators=10, random_state=42)

In [52]:
reg_model.feature_importances_

array([0.07865622, 0.0057302 , 0.00492833, 0.0059707 , 0.00612225,
       0.00814527, 0.01323693, 0.01362434, 0.01305349, 0.00402607,
       0.00244715, 0.00245281, 0.00350821, 0.00241576, 0.00409036,
       0.00250239, 0.00450296, 0.01070222, 0.00613171, 0.00581333,
       0.00962453, 0.009569  , 0.00955233, 0.01188935, 0.00678683,
       0.00788384, 0.01242929, 0.01228629, 0.01026648, 0.00765815,
       0.09694511, 0.07844439, 0.29166362, 0.17094956, 0.06599052])

In [48]:
regressor_step = ('regressor', RandomForestRegressor())

model_regressor_steps = [data_prep_step, regressor_step]

pipe_model = Pipeline(model_regressor_steps)

In [47]:
pipe_model.fit(X_train_trans, y_train_trans)

KeyError: "None of [Index(['Hotel_Address', 'Hotel_Name', 'Review_Date', 'Reviewer_Nationality',\n       'Tags'],\n      dtype='object')] are in the [columns]"

In [54]:
from sklearn.model_selection import GridSearch


ImportError: cannot import name 'GridSearch' from 'sklearn.model_selection' (/home/llbernat/anaconda3/envs/musi-aa/lib/python3.7/site-packages/sklearn/model_selection/__init__.py)

Seguim preparant per fer la nostra classificació ...

### 5. Entrena el modelo y revisa las variables/palabras que han sido más importantes.

In [35]:
## Separam la variable dependent

X = df.drop(['Reviewer_Score'], axis = 1)

## Classificador: si la puntuació és més gran que goodScore, aleshores la consideram bona

minGoodScore = 8.4

y = df['Reviewer_Score'] >= minGoodScore

## A veure que tenim ...
y.value_counts()

True     293974
False    221764
Name: Reviewer_Score, dtype: int64

### 6. Cuéntale al cliente un mensaje clave ... 
... sobre que tipo de cosas tienen más en cuenta los clientes: `Si no quieres una mala review no sirvas comida fría`